cannon/mipsevm/memory/memory.go

```diff
@@ -280,18 +280,22 @@ func (m *Memory) SetMemoryRange(addr Word, r io.Reader) error {
 	for {
 		pageIndex := addr >> PageAddrSize
 		pageAddr := addr & PageAddrMask
-		p, ok := m.pageLookup(pageIndex)
-		if !ok {
-			p = m.AllocPage(pageIndex)
-		}
-		p.InvalidateFull()
-		n, err := r.Read(p.Data[pageAddr:])
+		readLen := PageSize - pageAddr
+		chunk := make([]byte, readLen)
+		n, err := r.Read(chunk)
 		if err != nil {
 			if err == io.EOF {
 				return nil
 			}
 			return err
 		}
+
+		p, ok := m.pageLookup(pageIndex)
+		if !ok {
+			p = m.AllocPage(pageIndex)
+		}
+		p.InvalidateFull()
+		copy(p.Data[pageAddr:], chunk)
```

Comment:
Should this just copy over the actual number of bytes read?:
```suggestion
		copy(p.Data[pageAddr:pageAddr+n], chunk)
```
It looks like we're potentially copying over extra 0's.

In any case, we should probably have a test that calls `SetMemoryRange` on a page with non-zero bytes throughout so we can assert that only the expected bytes are modified.

Reply:
Yup. Will fix.